### PR TITLE
ACI relationship with new field subcategory & substitute

### PR DESCRIPTION
**Pull request by bijujoseph**
The core idea is to add minimal possible meta-data in measures and update the build script to enrich the existing measures. 

The requirement from our side are: 
  Given a measure, I need to know
   1.  what kind of measure it is (example "pre-attestation" or "registry performance" etc
   2.  which other measures can substitute the given measure. 

The approach here uses a new field `subCategory` and `substitutes`. If you prefer reusing `subCategoryId` instead of the new field `subCategory` do let me know.

### CODE REVIEW

**Comment by sungchoi** [comment]
Just a heads up that v0.0.40 has been published so you'll want to be at 41 for this PR.

File: measures/measures-data.xml [hunk 1]
```diff
@@ -93,6 +101,10 @@ AND (2) If requested, cooperated in good faith with ONC direct review of his or
     <measureSet>transition</measureSet>
     <isBonus>false</isBonus>
     <objective>electronicPrescribing</objective>
+    <reportingCategory>base</reportingCategory>
+    <substitutes>ACI_TRANS_EP_1</substitutes>
```
Comment (marimiyachi) [inline measures/measures-data.xml]: This should be singular `substitute`.

File: measures/measures-schema.yaml [hunk 2]
```diff
@@ -57,6 +57,9 @@ definitions:
     type: object
     properties:
       category: { enum: [aci] }
+      reportingCategory:
+        description: The reporting category of the ACI measure.
+        enum: [pre-attestation, base, registryBonus, performanceBonus, cehrtBonus]
```
Comment (marimiyachi) [inline measures/measures-schema.yaml]: Please use camelCase `preAttestation`.

File: measures/measures-schema.yaml [hunk 3]
```diff
@@ -71,6 +74,9 @@ definitions:
         description: If true, attesting to the measure will qualify the provider for ACI bonus points.
         type: boolean
         default: false
+      substitutes:
+        description: Identifiers of other ACI Measures that can be used instead of the current meausre.
```
Comment (marimiyachi) [inline measures/measures-schema.yaml]: Type `measure`

File: scripts/merge-measures-data.js [hunk 4]
```diff
@@ -114,5 +114,23 @@ function mergeQpp(qppJson) {
...
 
+  enrichACIMeasures(qppJson);
+
   return JSON.stringify(qppJson, null, 2);
 }
+
+
+
+function enrichACIMeasures(measures) {
+    var aciRelations = require('../util/aci-measure-relations.json');
```
Comment (marimiyachi) [inline scripts/merge-measures-data.js]: Nit: spaces are off here.

File: measures/measures-schema.yaml [hunk 5]
```diff
@@ -123,16 +129,6 @@ definitions:
...
-      ema:
-        description: Eligibility Measure Adjustment(EMA) cluster relationship for applicable the measure.
-        type: object
-        properties:
-          claimsSpecialitySet:
-            type: array
-            items: { $ref: #/definitions/specialityClusterSet}
-          registrySpecialitySet:
-            type: array
-            items: { $ref: #/definitions/specialityClusterSet}
```
Comment (kencheeto) [inline measures/measures-schema.yaml]: are these removals related?

File: scripts/merge-measures-data.js [hunk 6]
```diff
@@ -114,5 +114,29 @@ function mergeQpp(qppJson) {
...
+ * Will add extra metadata to ACI measures
+ * @param measures - the measures to enrich
+ */
+function enrichACIMeasures(measures) {
+
+  // load the relations from file
+  var aciRelations = require('../util/aci-measure-relations.json');
+
+  measures
+      .filter(m => m.category === 'aci')
```
Comment (kencheeto) [inline scripts/merge-measures-data.js]: 2 space indent please! also we generally prefer spelling out `measure` instead of using shortenings; for a smaller script like this spelling out helps legibility

File: scripts/merge-measures-data.js [hunk 7]
```diff
@@ -114,5 +114,29 @@ function mergeQpp(qppJson) {
...
+
+
+/**
+ * Will add extra metadata to ACI measures
+ * @param measures - the measures to enrich
+ */
+function enrichACIMeasures(measures) {
+
+  // load the relations from file
+  var aciRelations = require('../util/aci-measure-relations.json');
```
Comment (kencheeto) [inline scripts/merge-measures-data.js]: const

File: scripts/merge-measures-data.js [hunk 8]
```diff
@@ -114,5 +114,29 @@ function mergeQpp(qppJson) {
...
+
+  measures
+      .filter(m => m.category === 'aci')
+      .forEach(m => {
+          // find the relation and upgrade the measureSet and add substitute
+          var aciRelation = aciRelations[m.measureId];
```
Comment (kencheeto) [inline scripts/merge-measures-data.js]: const

File: test/measures/measures-data-spec.js [hunk 9]
```diff
@@ -21,6 +21,40 @@ describe('measures data json', function() {
         .filter(x => requiredAttestationIdsSet.has(x)));
       assert.equal(intersection.size, requiredAttestationIdsSet.size);
     });
+    it('should not have substitutes', () => {
```
Comment (kencheeto) [inline test/measures/measures-data-spec.js]: newline before this test case, and we prefer writing testcases without 'should' - so `it('does not have substitutes')` is just as legible

File: test/measures/measures-data-spec.js [hunk 10]
```diff
@@ -21,6 +21,40 @@ describe('measures data json', function() {
...
+      requiredAttestationIdsSet.forEach(measureId => {
+        var measure = measuresData.find(m => m.measureId == 'ACI_INFBLO_1');
+        assert.isTrue( _.isEmpty(measure.substitutes));
```
Comment (kencheeto) [inline test/measures/measures-data-spec.js]: `assert.isEmpty(measure.substitutes)`

File: test/measures/measures-data-spec.js [hunk 11]
```diff
@@ -21,6 +21,40 @@ describe('measures data json', function() {
...
+      })
+    });
+
```
Comment (kencheeto) [inline test/measures/measures-data-spec.js]: don't need this newline

File: test/measures/measures-data-spec.js [hunk 12]
```diff
@@ -21,6 +21,40 @@ describe('measures data json', function() {
...
+      measuresData
+        .filter(m => m.category === 'aci')
+        .forEach(m =>  {
+            generated[m.measureId] = {reportingCategory: m.reportingCategory, substitutes: m.substitutes};
+        });
+
+      assert.deepEqual(generated, actual);
+
+    });
+
```
Comment (kencheeto) [inline test/measures/measures-data-spec.js]: whitespace is pretty inconsistent here - could you make sure everything is 2-spaced with newlines between `it` test cases, const over var?

File: test/measures/measures-data-spec.js [hunk 13]
```diff
@@ -21,6 +21,40 @@ describe('measures data json', function() {
...
+      var measure = measuresData.find(m => m.measureId == 'ACI_PHCDRR_1');
+      assert.equal(measure.reportingCategory, 'performanceBonus');
+    });
+    it('ACI_TRANS_PHCDRR_2 should contain correct substitutes', () => {
+      var measure = measuresData.find(m => m.measureId == 'ACI_TRANS_PHCDRR_2');
+      assert.deepEqual(measure.substitutes, ['ACI_PHCDRR_2']);
+    });
+
+    it('should have proper metadata on all measures', () => {
+      var actual = require('../../util/aci-measure-relations.json');
```
Comment (kencheeto) [inline test/measures/measures-data-spec.js]: it's good practice to only require at the top of a file, in the spirit of how es6 import statements work

File: util/aci-measure-relations.json [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  "ACI_TRANS_PEA_2" : {"reportingCategory":"performanceBonus", "substitutes": ["ACI_CCTPE_1"]},
+  "ACI_TRANS_MR_1" : {"reportingCategory":"performanceBonus", "substitutes": ["ACI_HIE_3"]},
+
+  "ACI_PHCDRR_2" : {"reportingCategory":"registryBonus", "substitutes": ["ACI_TRANS_PHCDRR_2"]},
+  "ACI_PHCDRR_5" : {"reportingCategory":"registryBonus", "substitutes": ["ACI_TRANS_PHCDRR_3"]},
+  "ACI_PHCDRR_3" : {"reportingCategory":"registryBonus", "substitutes": []},
+  "ACI_PHCDRR_4" : {"reportingCategory":"registryBonus", "substitutes": []},
+  "ACI_TRANS_PHCDRR_2" : {"reportingCategory":"registryBonus", "substitutes": ["ACI_PHCDRR_2"]},
+  "ACI_TRANS_PHCDRR_3" : {"reportingCategory":"registryBonus", "substitutes": ["ACI_PHCDRR_5"]},
+
```
Comment (kencheeto) [inline util/aci-measure-relations.json]: are these newlines meaningful? if not, could we remove them?

File: scripts/merge-measures-data.js [hunk 15]
```diff
@@ -114,5 +114,29 @@ function mergeQpp(qppJson) {
...
   console.error('did not find measure details for the following', measuresNotFound);
 
+  enrichACIMeasures(qppJson);
+
   return JSON.stringify(qppJson, null, 2);
 }
+
+
+/**
+ * Will add extra metadata to ACI measures
```
Comment (kencheeto) [inline scripts/merge-measures-data.js]: Can you add a little more color here? What kind of metadata, where does the metadata come from (how is aci-measure-relations generated)?

**Comment by kencheeto** [comment]
Would it be simpler to merge in aci-measure-relations using jq?

File: measures/measures-schema.yaml (hunk 5 above)
Comment (bijujoseph) [inline measures/measures-schema.yaml]: Yes, those are debris left out from EMA cluster, which I forgot to remove.

File: scripts/merge-measures-data.js (hunk 6 above)
Comment (bijujoseph) [inline scripts/merge-measures-data.js]: Sure!, it is a good idea to create an [editor config file](http://editorconfig.org) so that formatting will be easier for internal devs and external contributors. I am adding one, configured it to have 2 spaces, but please adjust it to match other preferences later.

File: test/measures/measures-data-spec.js (hunk 9 above)
Comment (bijujoseph) [inline test/measures/measures-data-spec.js]: sure!

File: test/measures/measures-data-spec.js (hunk 12 above)
Comment (bijujoseph) [inline test/measures/measures-data-spec.js]: Sure I correct it. 
It is really hard to speculate stringent formatting and coding styles. I recommend you create lint file so that editors will automatically pick it up. Also, it will help external/internal contributrors

Comment (bijujoseph) [inline test/measures/measures-data-spec.js]: @kencheeto I have added an issue #53 please review it.

**Comment by bijujoseph** [comment]
@kencheeto looks like you are the `jq` superhero :-). Do you know if we can lookup by key (measureId) and merge in measure array? If yes, let me know, that is in short what I need, it will avoid me googling..

File: scripts/merge-measures-data.js (hunk 7 above)
Comment (bijujoseph) [inline scripts/merge-measures-data.js]: done! also moved it up to the top

File: scripts/merge-measures-data.js (hunk 15 above)
Comment (bijujoseph) [inline scripts/merge-measures-data.js]: Done!

File: test/measures/measures-data-spec.js (hunk 10 above)
Comment (bijujoseph) [inline test/measures/measures-data-spec.js]: No
```
TypeError: assert.isEmpty is not a function
    at test/measures/measures-data-spec.js:27:16
```

File: test/measures/measures-data-spec.js (hunk 11 above)
Comment (bijujoseph) [inline test/measures/measures-data-spec.js]: done!

File: test/measures/measures-data-spec.js (hunk 13 above)
Comment (bijujoseph) [inline test/measures/measures-data-spec.js]: Done!

File: test/measures/measures-data-spec.js (hunk 12 above)
Comment (kencheeto) [inline test/measures/measures-data-spec.js]: Agree! Glad to setup some tooling to help make this easier.

**Comment by lucasmbrown-usds** [comment]
Awesome collaborations, thanks all!

**Comment by bijujoseph** [comment]
@kencheeto please publish the module when you get a chance. Thank you

**Comment by marimiyachi** [comment]
@bijujoseph published!
